sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs

```diff
@@ -1863,44 +1862,32 @@ impl ty::TyExpression {
                     .by_ref()
                     .with_help_text("")
                     .with_type_annotation(type_engine.insert(engines, initial_type.clone(), None));
-                Self::type_check(handler, ctx, expr)
+
+                // type_check_analyze unification will give the final error
+                let handler = Handler::default();
+                Self::type_check(&handler, ctx, expr)
                     .unwrap_or_else(|err| ty::TyExpression::error(err, span, engines))
             })
             .collect();
 
-        // choose the best type to be the array elem type
-        use itertools::FoldWhile::{Continue, Done};
-        let elem_type = typed_contents
-            .iter()
-            .fold_while(None, |last, current| match last {
-                None => Continue(Some(current.return_type)),
-                Some(last) => {
-                    if last.is_concrete(engines, TreatNumericAs::Abstract) {
-                        return Done(Some(last));
-                    }
-
-                    if current
-                        .return_type
-                        .is_concrete(engines, TreatNumericAs::Abstract)
-                    {
-                        return Done(Some(current.return_type));
-                    }
-
-                    let last_info = ctx.engines().te().get(last);
-                    let current_info = ctx.engines().te().get(current.return_type);
-                    match (&*last_info, &*current_info) {
-                        (TypeInfo::Numeric, TypeInfo::UnsignedInteger(_)) => {
-                            Done(Some(current.return_type))
-                        }
-                        _ => Continue(Some(last)),
-                    }
-                }
-            })
-            .into_inner();
-        let elem_type = elem_type.unwrap_or_else(|| typed_contents[0].return_type);
+        // if the element type is still unknown, and all elements are Never,
+        // fallback to unit
+        let initial_type = if matches!(initial_type, TypeInfo::Unknown) {
```

Comment:
Is there a problem with using `Unknown` instead of `unit` as the result in this case?

If the type is still `Unknown` after unification, then we should throw an error because the type could not be determined. Otherwise it seems to me that using `Unknown` here solves the entire issue.

Reply:
But giving an error in this case does not seem wrong to me. We just need to decide which we prefer.

Reply:
I don't think it matters much so long as we guarantee that the unannotated type of `[]` is `[();0]`.

I suppose in the broadest sense, the programmer's intent with a program like this is unambiguous and so it's not really erroneous. It's just weird, but it's fully deterministic and since the elements aren't used at all, any zero sized type will do.

Reply:
I think that in this case ("arrays without typing context and all elements are never") we have four options:
1 - infer the array element as never; so something like `[!;1]`;
2 - infer the whole array as never;
3 - decay into a zero-sized type, in this case, `unit`;
4 - give an error.

`1` and `2` create this complicated case of how to deal with  "generic types with never". 
I think `4` is totally reasonable in this case.
But I implemented `3` just because it is what Rust does.